### ISSUE TITLE
Low timeout values causes duplicate commands to execute

### ISSUE THREAD

**Issue by vitaliyf** [saltstack/salt#32368]
### Description of Issue/Question

We are seeing that all commands when executed via ```sudo salt -G key:value foo.bar``` end up executing twice. This is reproduced by doing a ```cmd.run "date >>/tmp/date.txt"``` and observing two lines appended to that file, with dates about 10 seconds apart.

By observing ```salt-run state.event``` on master, we see that there are two jobs being sent and returned about 10 seconds apart.  Of course, only results of first one are shown on the CLI, so results of 2nd job just get ignored. Of course, it's a pretty dangerous this to happen since you may not be aware that something was executed twice.

We narrowed this down to "timeout" master setting.  We increased it from default 5 to 15 seconds and problems went away.

### Setup

Master is a dual CPU physical server that does nothing but salt-master, with a couple hundred minions.

### Steps to Reproduce Issue

From my debugging, I narrowed it down to grain matching taking a long time (longer than timeout) and then something kicking off 2nd identical job.  I did not get around to figure out what does that kick-off, you may be able to do that faster since you know the codebase.

Timeline looks roughly like this:

hh:mm:00 Run salt -G foo:bar test.ping
hh:mm:08 Event bus is showing job being sent to matched minions
hh:mm:09 Event bus shows results arriving back, and terminal window prints it, salt command exits
hh:mm:14 Event bus is showing brand new job (different id, same parameters) sent to same minions
hh:mm:15 Event bus showing results coming back

Between :00 and :08 I see CPU usage by a single salt-master thread peaked at 100%, and also master's debug logs showing grain-matching activity.

### Versions Report

Salt Version:
           Salt: 2015.8.8.2

Dependency Versions:
         Jinja2: 2.7.1
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.5.0
         Python: 2.7.10 (default, May 28 2015, 10:59:31)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: Not Installed
       cherrypy: 3.6.0
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.4
      pycparser: Not Installed
       pycrypto: 2.6
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist: centos 6.7 Final
        machine: x86_64
        release: 2.6.32-573.7.1.el6.x86_64
         system: CentOS 6.7 Final

**Comment by jfindlay** [comment]
@vitaliyf, thanks for debugging and reporting this.

**Comment by cachedout** [comment]
I can confirm this but _only_ by artificially inflating the time that it takes to match a grain.

```
diff --git a/salt/utils/minions.py b/salt/utils/minions.py
index 41858ab..182adf9 100644
--- a/salt/utils/minions.py
+++ b/salt/utils/minions.py
@@ -269,6 +269,8 @@ class CkMinions(object):
         '''
         Return the minions found by looking via grains
         '''
+        import time
+        time.sleep(9)
         return self._check_cache_minions(expr, delimiter, greedy, 'grains')
 
     def _check_grain_pcre_minions(self, expr, delimiter, greedy):
```

Under what I would consider normal operation, this is unlikely to happen. I'd consider this a bug but more of an edge case

Do you have any insight into what's causing your traversal of your master cache to take so long?

**Comment by cachedout** [comment]
So, all indications here seem to point to this being in the LocalClient but I haven't been able to find the source of this bug. We're definitely getting multiple publications to the master, so it's just a matter of tracking down where that's happening.

**Comment by vitaliyf** [comment]
Thanks for looking into this.

I'm not sure why grain-matching is taking this long.  This is a fairly beefy (couple dozen cores, 32gb ram) physical server that only does salt-master work for a couple hundred minions, so it's not underpowered.  We do have a handful of custom _grains/*.py on each minion, but nothing extravagant.  Across those couple hundred minions, there are:

```
$ sudo salt '*' grains.items |wc -l
62534
```

I was looking through Salt changelogs and commit history and recall finding some changes related to grain caching which prompted me to upgrade to latest 2018.8.8.2 release, but that didn't make any difference.

During those 8-9 seconds that are spend matching minions I can see one .py thread/process spiking up to 100% cpu usage, so it's definitely working hard.  When I set master's log level to debug, there is a ton of these lines:

```
Attempting to match 'something' in 'roles' using delimiter ':'
```

so perhaps the issue is that we have one /etc/salt/grain entry that is a list - but this "roles: [foo, bar]" is a very frequent pattern for people to use from what I've seen.

Of course, regardless why matching takes so long, republishing the same job is the real problem.

**Comment by DmitryKuzmenko** [comment]
These are PRs for

> When I set master's log level to debug, there is a ton of these lines:
> ```Attempting to match 'something' in 'roles' using delimiter ':'```

**Comment by DmitryKuzmenko** [comment]
The described duplicate commands issue is fixed by #31364 that is first time released in 2015.8.9 and included in all later 2015.8.x and 2016.x releases.
